kolibri/core/assets/src/heartbeat.js

```diff
@@ -15,68 +16,98 @@ import {
+    this._pollSessionEndPoint = this._pollSessionEndPoint.bind(this);
+    this.setUserInactive();
+    this._enabled = false;
+  }
+  startPolling() {
+    if (!this._enabled) {
+      logging.debug('Starting heartbeat');
+      this._enabled = true;
+      this._setActivityListeners();
+      // Do an immediate check to populate session info
+      return this._pollSessionEndPoint();
+    }
+    // Either return the promise for the current session endpoint
+    // poll, or return an immediately resolved promise, if that
+    // has already completed and been resolved.
+    return this._activePromise || Promise.resolve();
+  }
+  stopPolling() {
     logging.debug('Stopping heartbeat');
-    this.enabled = false;
-    this.clearActivityListeners();
-    if (this.timerId) {
-      clearTimeout(this.timerId);
+    this._enabled = false;
+    this._clearActivityListeners();
+    if (this._timerId) {
+      clearTimeout(this._timerId);
     }
   }
-  setActivityListeners() {
-    this.events.forEach(event => {
-      document.addEventListener(event, this.setActive, { capture: true, passive: true });
+  _setActivityListeners() {
+    this._userActivityEvents.forEach(event => {
+      document.addEventListener(event, this.setUserActive, { capture: true, passive: true });
     });
   }
-  clearActivityListeners() {
-    this.events.forEach(event => {
-      document.removeEventListener(event, this.setActive, { capture: true, passive: true });
+  _clearActivityListeners() {
+    this._userActivityEvents.forEach(event => {
+      document.removeEventListener(event, this.setUserActive, { capture: true, passive: true });
     });
   }
-  setActive() {
-    if (this.active !== true) {
-      this.active = true;
-      this.clearActivityListeners();
+  setUserActive() {
+    if (this._active !== true) {
+      this._active = true;
+      this._clearActivityListeners();
     }
   }
-  setInactive() {
-    this.active = false;
+  setUserInactive() {
+    this._active = false;
   }
-  wait() {
+  get _delay() {
     const { reconnectTime } = store.getters;
-    // If we are currently engaged in exponential backoff in trying to reconnect to the server
-    // use the current reconnect time preferentially instead of the standard delay.
-    // The reconnect time is stored in seconds, so multiply by 1000 to give the milliseconds.
-    this.timerId = setTimeout(this.beat, reconnectTime * 1000 || this.delay);
-    return this.timerId;
+    if (!store.getters.connected && reconnectTime) {
+      // If we are currently engaged in exponential backoff in trying to reconnect to the server
+      // use the current reconnect time preferentially instead of the standard delay.
+      // The reconnect time is stored in seconds, so multiply by 1000 to give the milliseconds.
+      return reconnectTime;
+    }
+    // If page is not visible, don't poll as frequently, as user activity is unlikely.
+    return store.state.pageVisible ? ACTIVE_DELAY : ACTIVE_DELAY * 2;
```

Comment:
is there any objetive reason to the x2 factor when the page is not visible, or it's just a guess?

Reply:
It's less frequent, but not so infrequent that it would cause a session timeout with our default setting.